api/python/t4/packages.py

```diff
@@ -113,21 +114,77 @@ def _clone(self):
         return PackageEntry(copy.deepcopy(self.physical_keys), self.size, \
                             copy.deepcopy(self.hash), copy.deepcopy(self.meta))
 
+
 class Package(object):
     """ In-memory representation of a package """
 
-    def __init__(self, data=None, meta=None):
+    def __init__(self, name=None, pkg_hash=None, registry=None):
         """
-        _data is of the form {logical_key: PackageEntry}
-        """
-        self._data = data or {}
-        self._meta = meta or {}
+        Create a Package from scratch, or load one from a registry.
+
+        Args:
+            name(string): name of package to load
+            pkg_hash(string): top hash of package version to load
+            registry(string): location of registry to load package from
+        """
+        if name is None and pkg_hash is None and registry is None:
+            self._data = {}
+            self._meta = {'version': 'v0'}
+            return
+
+        if registry is None:
+            # default to local registry
+            registry = get_package_registry()
+
+        if pkg_hash is not None:
+            # if hash is specified, name doesn't matter
+            pkg_path = '{}/packages/{}'.format(registry, pkg_hash)
+            # TODO replace open with something that supports both local and s3
+            self = self._from_path(pkg_path)
+            return
+
+        pkg_path = '{}/named_packages/{}/'.format(registry, quote(name))
+        latest = urlparse(pkg_path + 'latest')
+        if latest.scheme == 'file':
+            latest_path = unquote(latest.path)
+            with open(latest_path) as latest_file:
+                latest_hash = latest_file.read()
+        elif latest.scheme == 's3':
+            bucket, path, vid = parse_s3_url(latest)
+            latest_bytes = download_bytes(bucket + path)
+            latest_hash = latest_bytes.decode('utf-8')
+        else:
+            raise NotImplementedError
+
+        latest_hash = latest_hash.strip()
+        latest_path = '{}/packages/{}'.format(registry, quote(latest_hash))
+        self = self._from_path(latest_path)
+
+    @staticmethod
+    def _from_path(uri):
+        """ Takes a URI and returns a package loaded from that URI """
+        src_url = urlparse(uri)
+        if src_url.scheme == 'file':
+            with open(unquote(src_url.path)) as open_file:
```

Comment:
`url2pathname`, not `unquote`. Though I'm still figuring out how to do this properly :-\